### PR TITLE
add `iter` module to standard library

### PR DESCRIPTION
**Pull request by 1Kinoti**
# Description
 
this pr introduces an `iter` module
to the standard library.

the module is aimed at extending the filter commands.

### CODE REVIEW

**Comment by fdncred** [comment]
very cool work!

File: crates/nu-std/tests/test_iter.nu [hunk 1]
```diff
@@ -0,0 +1,55 @@
+use std *
+
+export def test_iter_cmd [] {
```
Comment (presidento) [inline crates/nu-std/tests/test_iter.nu:3]: Why do you need this? In most cases it is better to export the individual test commands.

File: crates/nu-std/tests/test_iter.nu [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    assert equal $res [a 5 b 5 c 5 d 5 e]
+
+    let res = (1..4 | iter intersperse 0)
+    assert equal $res [1 0 2 0 3 0 4]
+
+    let res = (4 | iter intersperse 1)
+    assert equal $res [4]
+}
+
+def test_iter_scan [] {
```
Comment (presidento) [inline crates/nu-std/tests/test_iter.nu:46]: It seems that this command is not called. I suggest exporting it to be called, or removing it if it is unnecessary.

**Review by presidento — CHANGES_REQUESTED** [review]

**Review by presidento — COMMENTED** [review]
OK from my side, thanks.

**Review by presidento — APPROVED** [review]

File: crates/nu-std/lib/iter.nu [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    | reduce -f [] {|it, acc|
+        if ($it.index == $len - 1) {
+           $acc ++ [$it.item]
+        } else {
+            $acc ++ [$it.item, $separator]
+        }
+    }
+}
+
+# Returns the a list of intermediate steps performed by `reduce`
```
Comment (amtoine) [inline crates/nu-std/lib/iter.nu:72]: there is a word that should not be here, right? :yum:
"*Returns **the a** list*" :wink:

File: crates/nu-std/lib/iter.nu [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+#
+# assert equal $scanned [0, 1, 3, 6]
+#
+# # use the --noinit(-n) flag to remove the initial value from
+# # the final result
+# let scanned = ([1 2 3] | scan 0 {|x, y| $x + $y} -n)
+#
+# assert equal $scanned [1, 3, 6]
+# ```
+export def "iter scan" [ # -> list<any>
```
Comment (amtoine) [inline crates/nu-std/lib/iter.nu:91]: i'm not sure i would have guessed what this command does without the examples :thinking: 
what about `iter accumulate`? :yum:

**Review by amtoine — COMMENTED** [review]
these are quite cool actually :clap: :clap: 

i have a single question and that will be great to land :rocket:

File: crates/nu-std/lib/iter.nu (hunk 4 above)
Comment (presidento) [reply crates/nu-std/lib/iter.nu:91]: I vote for common names.

Comment (bobhy) [reply crates/nu-std/lib/iter.nu:91]: I think good names are the hardest part of programming, so pardon me for belaboring a point here.

I'm OK with Nu inviting analogies with Rust, especially if that's the primary language that it borrows from.  So I'd eventually get calling this function `scan` (though I'd be annoyed if it turns out to work subtly differently in Nushell).  `Accumulate` is a term used in Rust too, though not for a verb in the API, so I'd be less happy about that choice.

But roping Haskell into this discussion seems to be a stretch.  I don't know Haskell and all I can find with a quick search is a Haskell `scanl` and a `scanr`, that doesn't clarify anything for me.

Comment (amtoine) [reply crates/nu-std/lib/iter.nu:91]: @1Kinoti
thanks for the explaination, i just did not know it was a common name :wink: 

that's perfectly fine with me then :relieved:

**Review by amtoine — APPROVED** [review]
thanks @1Kinoti :sparkles:

**Comment by amtoine** [comment]
yeaaaaah it did land :partying_face: 

> looking forward to more work on std @amtoine
>
definitely, when you want :wink:

**Pull request merged**